src/dodal/devices/focusing_mirror.py

```diff
@@ -76,21 +76,26 @@ async def set(self, value, *args, **kwargs):
 
         LOGGER.debug(f"setting {setpoint_v.name} to {value}")
 
-        it_values = observe_value(demand_accepted, timeout=DEFAULT_SETTLE_TIME_S)
-        accepted_value = await anext(it_values)
-        if accepted_value == MirrorVoltageDemand.OK:
-            await setpoint_v.set(value)
-        else:
-            raise AssertionError(
-                f"Demand not accepted {accepted_value} before set attempted"
-            )
-
-        accepted_value = await anext(it_values)
+        # Register an observer up front to ensure we don't miss events after we
+        # perform the set
+        demand_accepted_iterator = observe_value(
+            demand_accepted, timeout=DEFAULT_SETTLE_TIME_S
+        )
+        # discard the current value (OK) so we can await a subsequent change
+        await anext(demand_accepted_iterator)
+        await setpoint_v.set(value)
+
+        # The set should always change to SLEW regardless of whether we are
+        # already at the set point, then change back to OK/FAIL depending on
+        # success
+        accepted_value = await anext(demand_accepted_iterator)
```

Comment:
Should: Sorry, I'm wondering if there is another race condition here where if we get an update straight after the set we may still see an OK? I think a better solution might be not to discard the value before the set but to keep reading until the updates show `slew`?

Reply:
I did look at the code in ophyd_async before to confirm that we will always get the current value in the first event, so I don't think that there is any possibility of this race condition and we should be safe.

`observe_value`'s implementation calls `signal.subscribe_value(q.put_nowait)`
before awaiting new values.
`signal.subscribe()` calls `signal_notify()` which calls our `q.put_nowait()` - this posts the event on to the queue which is then collected by the subsequent `await`.

If this implementation ever changes it will probably break calling code as anything that follows this model under this assumption will end up blocking waiting for an event that never comes.

I'm also slightly confused where you think the extra event will come from - we've already validated that the demand will be `OK` prior to entering the function and we should be the only ones applying changes. The discard from the first `anext()` is before the set, so there is no way that the discarded value can be "SLEW" or a subsequent "OK", even if the current value wasn't pushed by the call to `observe`.

Reply:
Tom's comment https://github.com/DiamondLightSource/dodal/pull/636#discussion_r1649079687 implies that it is possible we get multiple updates with the same value. It depends quite a bit on how the IOC is written, for example it's possible that the PV will "send out" an update on a periodic basis regardless of if the value has changed or not. I think you're right that I'm being overly cautious though, looking at the PV it seems that it should be stable.